runtime/interpreter/value_accountcapabilitycontroller.go

```diff
@@ -278,29 +284,32 @@ func (v *AccountCapabilityControllerValue) ReferenceValue(
 	)
 }
 
-// SetDeleted sets the controller as deleted, i.e. functions panic from now on
-func (v *AccountCapabilityControllerValue) SetDeleted(gauge common.MemoryGauge) {
-
-	raiseError := func() {
+// checkDeleted checks if the controller is deleted,
+// and panics if it is.
+func (v *AccountCapabilityControllerValue) checkDeleted() {
+	if v.deleted {
 		panic(errors.NewDefaultUserError("controller is deleted"))
 	}
+}
 
-	v.SetTag = func(s *StringValue) {
-		raiseError()
-	}
-	v.GetTag = func() *StringValue {
-		raiseError()
-		return nil
-	}
+func (v *AccountCapabilityControllerValue) newDeleteFunction(
+	inter *Interpreter,
+) *HostFunctionValue {
+	return NewHostFunctionValue(
+		inter,
+		sema.AccountCapabilityControllerTypeDeleteFunctionType,
+		func(invocation Invocation) Value {
+			// NOTE: check if controller is already deleted
+			v.checkDeleted()
```

Comment:
It seems very fragile to require that `v.checkDeleted()` is called in every new function added to a CapabilityController like this. Is it possible instead to wrap these functions in a wrapper that checks whether the controller is deleted in `GetMember`, instead of requiring it to be handled here in each function? I am concerned about how easy it is to forget to do this by accident.

Reply:
All I can think of is a helper like e.g. `func (StorageCapabilityControllerValue) newHostFunctionValue(...)` which calls `NewHostFunctionValue` with a function that calls `checkDeleted` at the beginning. But then we have to make sure we call `newHostFunctionValue` instead of `NewHostFunctionValue`, so didn't really gain much. It would also be harder to see a difference between a normal host function construction vs one that uses the helper – currently the calls to `checkDeleted` are explicit, but at least it's easier to see when it is missing, when comparing to other function bodies.

Did you have a particular idea how we could make this less error prone? I'd love to, but don't really have any ideas that would make it much better.

Reply:
What if we made `newHostFunctionValue` return a tuple of `Value` and some special other type (`DeletionChecked`?) that is only created by the helper. We could then have all of these other functions like `newDeleteFunction` and `newTargetFunction` return this same tuple, and then enforce in `GetMember` and `SetMember` that these functions do return this `DeletionChecked` that serves as "proof" that the special `newHostFunctionValue` was actually called. This would enforce at the type system level that this function is called, but on the other hand perhaps it is getting slightly overcomplicated?

Reply:
If we still need to ensure a helper is used (`newHostFunctionValue`), then we can maybe just check the deletion there, instead of requiring the implementation (the function passed to `newHostFunctionValue`) to perform the deletion check?

Reply:
Added a run-time based check in 4957cb11689d80c2459085ea58f28d897a536057. How does that look?

Reply:
Nice! That makes me feel a lot better about this.